### PR TITLE
Some queries for runtime fields

### PR DESCRIPTION
**Pull request by nik9000**
This implements a few queries for runtime fields. It doesn't wire them into the
runtime fields infrastructure yet because, well, we haven't committed that. But
the queries are there and they work. As do the doc values implementations.

You may wonder, Nik, why are you reimplementing all of these queries? Lucene
has tons of queries, some of them even against doc values. You make doc values,
why not just query the doc values you make with the Lucene queries? That'd be
a good question! It turns out that Lucene's doc_values queries are quite tied
up with Lucene's doc values implementations. They compare values using the
binary format that lucene stores the doc values in. Which is a great idea! But
it means that they only really work on Lucene's doc values implementations.

It also tries to only call each script once, even asserting that in most tests.
This isn't strictly *required* but I think it is useful and something that
would be difficult to design "after the fact". Still, this behavior shouldn't
block implementing the feature. We'll keep it if we can. But if we can't, well,
we'll drop and it and come back to it later.

relates to #59332

### CODE REVIEW

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractRuntimeValues.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                            @Override
+                            public boolean matches() throws IOException {
+                                leafCursor.accept(approximation.docID());
+                                return AbstractRuntimeQuery.this.matches();
+                            }
+
+                            @Override
+                            public float matchCost() {
+                                // TODO we have no idea what this should be and no real way to get one
+                                return 1000f;
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractRuntimeValues.java:139]: I think the cost can be for now the approximation.cost() , it is proportional to maxDocs if I remember correctly

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeValues.java [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+        protected void sort() {
+            Arrays.sort(values, 0, count);
+        }
+
+        private CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+            alwaysSortResults();
+            return DocValues::new;
+        }
+
+        private class DocValues extends SortedBinaryDocValues {
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeValues.java:116]: this is effectively the docvalues implementation that replaces the one that I had written which you deleted?

I was wondering, and not even sure if it matters, queries don't need sorted doc_values right? But we always implement the interface that we need for fielddata to simplify things? I know that I had implemented the lucene one for query but it made things more complex for almost no reason, I think.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractRuntimeValues.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                        TwoPhaseIterator twoPhase = new TwoPhaseIterator(approximation) {
+                            @Override
+                            public boolean matches() throws IOException {
+                                leafCursor.accept(approximation.docID());
+                                return AbstractRuntimeQuery.this.matches();
+                            }
+
+                            @Override
+                            public float matchCost() {
+                                // TODO we don't have a good way of estimating the complexity of the script so we just go with 9000
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractRuntimeValues.java:138]: Looking at the javadoc, I *think* it should actually just be some constant - `matchCost` says it is only for a single document.

If I just returned `1` here then I think Lucene'd reason about the query as though it were `match_all`, which is probably bad

File: server/src/main/java/org/elasticsearch/search/lookup/LeafDocLookup.java [hunk 4]
```diff
@@ -75,7 +83,12 @@ public void setDocument(int docId) {
             scriptValues = AccessController.doPrivileged(new PrivilegedAction<ScriptDocValues<?>>() {
                 @Override
                 public ScriptDocValues<?> run() {
-                    return fieldDataLookup.apply(fieldType).load(reader).getScriptValues();
+                    // TODO should this go through QueryShardContext?
+                    IndexFieldData<?> ifd = fieldDataLookup.apply(fieldType);
+                    if (ifd instanceof SearchLookupAware) {
+                        ((SearchLookupAware) ifd).setSearchLookup(searchLookup);
+                    }
+                    return ifd.load(reader).getScriptValues();
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafDocLookup.java:91]: Yeah, that is what my TODO was about - maybe this should go through `getForField`. I can look into making that change in master. I think it is good regardless. And, if it has consequences that we didn't think of we may as well see them earlier rather than later...

File: server/src/main/java/org/elasticsearch/index/fielddata/SearchLookupAware.java [hunk 5]
```diff
@@ -21,6 +21,8 @@
 
 import org.elasticsearch.search.lookup.SearchLookup;
 
+import java.io.IOException;
+
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/fielddata/SearchLookupAware.java:25]: :+1:

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeValues.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeValues.java:116]: > this is effectively the docvalues implementation that replaces the one that I had written which you deleted?

Yeah. I had mine sitting around and it had unit tests and stuff. Its the same deal.

> I was wondering, and not even sure if it matters, queries don't need sorted doc_values right? But we always implement the interface that we need for fielddata to simplify things? I know that I had implemented the lucene one for query but it made things more complex for almost no reason, I think.

In the patch as it stands now the queries don't use this and look at the array of values directly. And they don't sort them unless they build this thing. When I was writing the queries it felt simpler not to target the doc values implementation - partially because most queries don't care about the values being sorted. And, if they *did* care about the values being sorted, then they'd want a sorted array. Which is what we use to build the doc values implementation anyway.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java [hunk 6]
```diff
@@ -60,20 +68,78 @@ public String typeName() {
...
 
+    private StringRuntimeValues runtimeValues(QueryShardContext ctx) {
+        // TODO cache the runtimeValues in the context somehow
+        return scriptFactory.newFactory(script.getParams(), ctx.lookup()).runtimeValues();
+    }
+
     @Override
     public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
-        // TODO once we get SearchLookup as an argument, we can already call scriptFactory.newFactory here and pass through the result
+        // TODO figure out how to use runtimeValues here
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:78]: Its sort of the same TODO. To use `runtimeValues` we'd need the factory, is what I mean. i'll restore your TODO so it is more clear.

File: server/src/main/java/org/elasticsearch/search/lookup/LeafDocLookup.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafDocLookup.java:91]: I don't know if this makes sense on master and what difference it would make, but I think it does make sense in our feature branch to isolate the hack around augmenting the fielddata impl.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+            if (values.length < newCount) {
+                values = Arrays.copyOf(values, ArrayUtil.oversize(newCount, RamUsageEstimator.NUM_BYTES_OBJECT_REF));
+            }
+            values[count] = value;
+            count = newCount;
+        }
+
+        @Override
+        protected void sort() {
+            Arrays.sort(values, 0, count);
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java:103]: sorting is always done in the same way regardless of the data type right? I mean calling Arrays.sort

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+    public Query termsQuery(String fieldName, String... values) {
+        return new Values().new TermsQuery(fieldName, values);
+    }
+
+    public Query wildcardQuery(String fieldName, String pattern) {
+        return new Values().new WildcardQuery(fieldName, pattern);
+    }
+
+    private class Values extends AbstractRuntimeValues {
+        private String[] values = new String[1];
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java:82]: should we just have the collection of values shared between the different types and use a typed ArrayList? Would that be a tragedy in terms of performance?

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+        protected void sort() {
+            Arrays.sort(values, 0, count);
+        }
+
+        private CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+            alwaysSortResults();
+            return DocValues::new;
+        }
+
+        private class DocValues extends SortedBinaryDocValues {
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java:111]: like I said in other comments, I think that you have all the bits that are needed, I only would turn around a bit the dependencies/flow. I would keep the existing doc_values impl, which is only used in aggs/sorting (and it is totally fine that it's not used in queries), and I would make it lighter so that all it does is really returning the result of the script for the current doc, by calling that script.getResult(docId) method I mentioned above. I think that would clarify who does what. At the moment it is hard to follow who needs doc_values who needs them sorted etc.

**Comment by javanna** [comment]
shall we close this @nik9000 ? I think that it was a good experiment and we can now proceed getting in what we need step by step.

**Comment by nik9000** [comment]
We got this in with #59630, #59527, and #59523.
